resources/aws_ec2.go

```diff
@@ -0,0 +1,535 @@
+	}
+	obj.client = ec2.New(sess)
+
+	obj.wg = &sync.WaitGroup{}
+
+	return obj.BaseRes.Init() // call base init, b/c we're overriding
+}
+
+// Watch is the primary listener for this resource and it outputs events.
+func (obj *AwsEc2Res) Watch() error {
+	send := false
+	var exit *error
+	if err := obj.Running(); err != nil {
+		return err
+	}
+	type chanStruct struct {
+		str string
+		err error
+	}
+	awsChan := make(chan *chanStruct)
+	closeChan := make(chan struct{})
+	ctx, cancel := context.WithCancel(context.TODO())
+	obj.wg.Add(1)
+	go func() {
+		defer obj.wg.Done()
+		select {
+		case <-closeChan:
+			cancel()
+		}
+	}()
+	obj.wg.Add(1)
+	defer obj.wg.Wait()
+	go func() {
+		defer obj.wg.Done()
+		defer close(awsChan)
+		for {
+			diInput := &ec2.DescribeInstancesInput{
+				Filters: []*ec2.Filter{
+					{
+						Name:   aws.String("tag:Name"),
+						Values: []*string{aws.String(obj.prependName())},
+					},
+					{
+						Name: aws.String("instance-state-name"),
+						Values: []*string{
+							aws.String("pending"),
+							aws.String("running"),
+							aws.String("stopping"),
+							aws.String("stopped"),
+						},
+					},
+				},
+			}
+			diOutput, err := obj.client.DescribeInstances(diInput)
+			if err != nil {
+				awsChan <- &chanStruct{
+					err: errwrap.Wrapf(err, "error describing instances"),
+				}
+			}
+			if obj.State == "running" {
+				stoppedInput := &ec2.DescribeInstancesInput{
+					Filters: []*ec2.Filter{
+						{
+							Name:   aws.String("tag:Name"),
+							Values: []*string{aws.String(obj.prependName())},
+						},
+						{
+							Name: aws.String("instance-state-name"),
+							Values: []*string{
+								aws.String("stopped"),
+							},
+						},
+					},
+				}
+				stoppedOutput, err := obj.client.DescribeInstances(stoppedInput)
+				if err != nil {
+					awsChan <- &chanStruct{
+						err: errwrap.Wrapf(err, "error describing instances"),
+					}
+				}
+				if len(diOutput.Reservations) == 1 && len(stoppedOutput.Reservations) == 0 {
+					waitInput := &ec2.DescribeInstancesInput{
+						InstanceIds: []*string{diOutput.Reservations[0].Instances[0].InstanceId},
+						Filters: []*ec2.Filter{
+							{
+								Name:   aws.String("instance-state-name"),
+								Values: []*string{aws.String("stopped"), aws.String("terminated")},
+							},
+						},
+					}
+					log.Printf("watching: %s", *diOutput.Reservations[0].Instances[0].InstanceId)
+					obj.client.WaitUntilInstanceStoppedWithContext(ctx, waitInput)
+					stateOutput, err := obj.client.DescribeInstances(diInput)
+					if err != nil {
+						awsChan <- &chanStruct{
+							err: errwrap.Wrapf(err, "error describing instances"),
+						}
+					}
+					var stateName string
+					if len(stateOutput.Reservations) == 1 {
+						stateName = *stateOutput.Reservations[0].Instances[0].State.Name
+					}
+					if len(stateOutput.Reservations) == 0 || (len(stateOutput.Reservations) == 1 && stateName != "running") {
+						x := &chanStruct{
```

Comment:
good, but you can remove the x temp variable and directly use the struct now that it's more clear what the syntax is